### ISSUE TITLE
TES3MP license changed, restrictions added, does not match upstream OpenMW license

### ISSUE THREAD

**Issue by backlabs1** [TES3MP/TES3MP#685]
Hi,

The upstream OpenMW repository uses the GPLv3 license without any [section 7](https://www.gnu.org/licenses/gpl-3.0.en.html#section7) additional terms. The TES3MP project forked the OpenMW repository and then later (commits f6887559f6baa2aa94847ee8341045ffeffdac8f and cf4ff3513f25574c6c79377c09fe72e95c2ee069) changed the license to GPLv3 with additional section 7 restrictions.

If I understand correctly:

- a GPLv3 licensee much have permission from all copyright holders to change the license of the forked codebase,
- merging in changes from upstream OpenMW requires permission from the copyright holders to re-license those changes, and
- using a different license for TES3MP prevents TES3MP changes from being merged to upstream OpenMW (which I understand to be a goal of this project) unless all corresponding TES3MP copyright holders agree to a license change.

(Note, a Git commit author is usually the copyright holder of the corresponding changes, so "copyright holders" above can in most cases be read as "commit authors".)

I assume this license change was a mistake or due to a misunderstanding (the GPLv3 section 7 details can be confusing). Or perhaps I am misinformed regarding the legal implications. I am opening this issue for discussion.

Thank you,
backlabs1

**Comment by davidcernat** [comment]
Please read the GPLv3 license again. Here's the wording it uses for terms that can be added to it:

> Notwithstanding any other provision of this License, **for material you add to a covered work**, you may (if authorized by the copyright holders of **that material**) supplement the terms of this License with terms:

In this case, we added additional terms for the code Koncord and I had contributed that constituted TES3MP. Note that it says "if authorized by the copyright holders of **that material**" (that material being TES3MP), _not_ "if authorized by the copyright holders of **that covered work**" (OpenMW).

> merging in changes from upstream OpenMW requires permission from the copyright holders to re-license those changes,

No, it does not, because that's still the OpenMW source code, not the TES3MP source code.

> using a different license for TES3MP prevents TES3MP changes from being merged to upstream OpenMW (which I understand to be a goal of this project) unless all corresponding TES3MP copyright holders agree to a license change.

In this case, the OpenMW license would either have to include the additional terms related to TES3MP — and there is **already** a precedent of OpenMW including additional terms when taking in [cc9cii](https://github.com/cc9cii)'s work, i.e. in files from openmw/components/esm4 — or Koncord and I would have to agree to switch the license back to the regular GPLv3, which isn't particularly hard to do.

**Comment by backlabs1** [comment]
Hi, and thank you for your reply.

>>**for material you add to a covered work**
>
>we added additional terms for the code Koncord and I had contributed that constituted TES3MP

I understand that, but the license was changed for everything—there is no copy of the original license. For example, the `/apps/niftest/*` files now appear to be covered by the TES3MP license although they have not been changed since the fork. That is a simple example, but the matter becomes much less clear with all the partial changes to OpenMW code, where those files are now marked as entirely TES3MP-licensed but are actually a mix of two licenses.

Also, if I contribute a change *anywhere* in this repository, without explicitly specifying the license in the commit message, it will be governed by the TES3MP license and will not be merge-able upstream without somehow adding an additional license to the individual line changes (but without dual-licensing because that would require re-licensing permission from previous contributors).

>the OpenMW license would either have to include the additional terms related to TES3MP

Which might be acceptable for TES3MP-only files, but difficult or impossible for changes to files that already exist upstream, because the TES3MP changes are TES3MP-licensed.

>or Koncord and I would have to agree to switch the license back to the regular GPLv3, which isn't particularly hard to do.

Would you consider doing that in order to eliminate any question on the matter?

Thank you,
backlabs1

**Comment by davidcernat** [comment]
> I understand that, but the license was changed for everything—there is no copy of the original license.

No, it was not changed for everything. The additional terms specifically apply to the "TES3MP GPL Source Code", which is the source code added by TES3MP.

> the matter becomes much less clear with all the partial changes to OpenMW code, where those files are now marked as entirely TES3MP-licensed but are actually a mix of two licenses.

They're almost always marked with "Start of TES3MP addition" or "Start of TES3MP change". You can also compare the source code between the two projects by yourself to see exactly which lines are TES3MP and which lines are not.

> Which might be acceptable for TES3MP-only files, but difficult or impossible for changes to files that already exist upstream, because the TES3MP changes are TES3MP-licensed.

It's not difficult. If the OpenMW included the additional terms in reference to the "TES3MP GPL Source Code", it would immediately solve this.

> Would you consider doing that in order to eliminate any question on the matter?

Not at this time. The license changes were made to address the problem of people making fake versions of TES3MP. Simply put, a multiplayer project really benefits from the additional GPLv3 terms compared to a singleplayer one.

**Comment by backlabs1** [comment]
Hi again,

>No, it was not changed for everything. The additional terms specifically apply to the "TES3MP GPL Source Code", which is the source code added by TES3MP.

That is not actually stated anywhere. In the context of the TES3MP repository, the "TES3MP GPL Source Code" refers to all the files.

>[The changes are] almost always marked with "Start of TES3MP addition" or "Start of TES3MP change".

But nothing is mentioned about the enclosed line(s) being under a different license.

>You can also compare the source code between the two projects by yourself to see exactly which lines are TES3MP and which lines are not.

So any added lines returned by `git diff openmw/openmw-0.47.0..tes3mp/0.8.1` are TES3MP-licensed? I'm not sure that works in all cases, but regardless, it is not specified anywhere. Even if it were, that would require the user to know which tag to compare with at any point in time and the third-party OpenMW repository and tag would need to still exist and be correct—all to find out what license governs a specific line in *this* repository.

**The licensing needs to be specified as described in [GPLv3 section 7](https://www.gnu.org/licenses/gpl-3.0.en.html#section7):**

>***If you add terms to a covered work in accord with this section, you must place, in the relevant source files, a statement of the additional terms that apply to those files, or a notice indicating where to find the applicable terms.***

Currently the TES3MP-only files[1] do not mention anything at all about any license, nor do they have the start/end markers.

As a possible resolution, I suggest creating a separate TES3MP license file, reverting the root LICENSE, and adding a statement **in each TES3MP-only file**[1] referencing the TES3MP license. That would eliminate any question about what falls under which license and it would make it clear that code from OpenMW is still under the same license as upstream.

If the TES3MP copyright holders in `git shortlog -se 7be09078b4..f6887559f6` (from fork to license change) have not already agreed to the license change, then that would need to be addressed first.

[1]: By "TES3MP-only files", I am referring to the files in the `/apps/browser`, `/apps/master`, `/apps/openmw-mp`, `components/openmw-mp`, and `files/tes3mp` directories. There are other files too that do not exist upstream.

Thanks again,
backlabs1
